plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -338,6 +363,76 @@ public IcebergTableHandle getTableHandle(
+        TableOperations ops = ((HasTableOperations) table).operations();
+        TableMetadata currentTableMetadata = ops.current();
+        metadataFileLocations.add(currentTableMetadata.metadataFileLocation());
+
+        collectSnapshotTableMetadatas(currentTableMetadata, snapshotId, snapshotTableMetadatas, metadataFileLocations, visitedMetadataFileLocations, ops.io());
+
+        if (snapshotTableMetadatas.isEmpty()) {
+            return Optional.empty();
+        }
+        else if (snapshotTableMetadatas.size() == 1) {
+            return Optional.of(snapshotTableMetadatas.get(0));
+        }
+        else {
+            List<String> snapshotTableMetadataFiles = snapshotTableMetadatas.stream().map(TableMetadata::metadataFileLocation).collect(Collectors.toList());
+
+            return snapshotTableMetadatas.stream()
+                    .filter(metadata -> metadata.previousFiles() == null ||
+                            metadata.previousFiles().isEmpty() ||
+                            // If there are multiple table metadata entries with the same snapshot,
+                            // pick the one that does not have any predecessors having the same snapshot
+                            metadata.previousFiles().stream().map(TableMetadata.MetadataLogEntry::file).noneMatch(snapshotTableMetadataFiles::contains))
+                    // In the unlikely case that two metadata entries have the same snapshot id and no predecessors with the
+                    // same snapshot id, take the oldest one.
+                    .min(Comparator.comparing(TableMetadata::lastUpdatedMillis));
+        }
+    }
+
+    private static void collectSnapshotTableMetadatas(
+            TableMetadata metadata,
+            long snapshotId,
+            List<TableMetadata> snapshotTableMetadatas,
+            Set<String> metadataFileLocations,
+            Set<String> visitedMetadataFileLocations,
+            FileIO io)
+    {
+        visitedMetadataFileLocations.add(metadata.metadataFileLocation());
+        if (metadata.currentSnapshot() != null && metadata.currentSnapshot().snapshotId() == snapshotId) {
+            snapshotTableMetadatas.add(metadata);
+        }
+
+        List<TableMetadata.MetadataLogEntry> metadataLogEntries = metadata.previousFiles();
+        if (metadataLogEntries.size() > 0) {
+            for (TableMetadata.MetadataLogEntry metadataLogEntry : metadataLogEntries) {
+                metadataFileLocations.add(metadataLogEntry.file());
+            }
+            for (TableMetadata.MetadataLogEntry metadataLogEntry : metadataLogEntries) {
+                String metadataLogEntryFile = metadataLogEntry.file();
+                if (visitedMetadataFileLocations.contains(metadataLogEntryFile)) {
+                    continue;
+                }
+                TableMetadata previousMetadata = null;
+                try {
+                    previousMetadata = TableMetadataParser.read(io, metadataLogEntryFile);
```

Comment:
How about adding a test case in `TestIcebergMetadataFileOperations` (optional `TestIcebergMetastoreAccessOperations`) so that we can track the access count during time travel?

Reply:
Added a new set of tests in the context of reading from a versioned table.